ui-spacetimechart/src/utils/canvas.ts

```diff
@@ -2,54 +2,89 @@ import { clamp, identity } from 'lodash';
 
 import { type PathEnd, type Point, type RGBAColor, type RGBColor } from '../lib/types';
 
+/**
+ * This function returns the picking layers scaling ratio. We basically take the min of the screen
+ * pixels and the "HTML pixels", and divide it by two.
+ *
+ * This allows having a smaller picking stage to fill (so it's faster), while keeping a "good enough
+ * precision".
+ */
+export function getPickingScalingRatio(): number {
+  const PICKING_DOWNSCALING_RATIO = 0.5;
+  const dpr = window.devicePixelRatio || 1;
+
+  // When devicePixelRatio is over 1 (like for Retina displays), we downscale based on the "HTML
+  // pixels":
+  if (dpr > 1) return PICKING_DOWNSCALING_RATIO;
+
+  // When devicePixelRatio is under 1 (like when the user zooms out for instance), we downscale
+  // based on the actual "screen pixels" (to avoid having a too large scene to fill):
+  if (dpr < 1) return PICKING_DOWNSCALING_RATIO * dpr;
+
+  return PICKING_DOWNSCALING_RATIO;
+}
+
 /**
  * This function draws a thick lines from "from" to "to" on the given ImageData, with no
  * antialiasing. This is very useful to handle picking, since it is not possible to disable
  * antialiasing with the native JavaScript canvas APIs.
  */
 export function drawAliasedLine(
   imageData: ImageData,
-  from: Point,
-  to: Point,
+  { x: fromX, y: fromY }: Point,
+  { x: toX, y: toY }: Point,
   [r, g, b]: RGBColor | RGBAColor,
   thickness: number,
   drawOnBottom: boolean,
-  number: number = Math.ceil(thickness / 2)
+  scalingRatio = 1
 ): void {
-  if (from.x > to.x)
-    return drawAliasedLine(imageData, to, from, [r, g, b], thickness, drawOnBottom);
+  if (fromX > toX)
+    return drawAliasedLine(
+      imageData,
+      { x: toX, y: toY },
+      { x: fromX, y: fromY },
+      [r, g, b],
+      thickness,
+      drawOnBottom
+    );
```

Comment:
We should probably pass down the `scalingRatio` here?

Reply:
Oh good catch, thanks! Fixed :+1: